drivers: usb: udc_dwc2: Support Remote Wakeup

tmon-nordic opened the pull request:
Rework bottom-half processing to improve hibernation handling necessary for proper remote wakeup from hibernation handling.

tmon-nordic committed:
drivers: usb: udc_dwc2: Replace queue with events

When the queue is full, all messages posted inside interrupt handlers
are simply dropped. This problem can be remedied by having the message
queue large enough, but determining the maximum number of messages that
can ever be posted in the system is really complex task.

Hopefully in DWC2 driver there is finite number of events that have to
be processed inside thread handler. Therefore it is unnecessary to
determine the maximum queue size for the events if the events are posted
to k_event object instead of send to k_msgq object.

Use combination of three k_event structures to handle all possible event
sources. This not only guarantees by design that no event will be lost,
but also slightly reduces the memory usage.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
drivers: usb: udc_dwc2: Synchronize thread with UDC API

UDC API handlers and DWC2 driver thread share not only software
constructs, but also the underlying hardware. Ensure that any UDC API
call is not preempted by DWC2 driver thread (and vice versa) by
acquiring the lock in thread handler.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
drivers: usb: udc_dwc2: Support Remote Wakeup

According to USB 2.0 Specification the remote wakeup device must hold
the resume signaling for at least 1 ms but for no more than 15 ms. The
DWC2 otg controller requires the software to drive the remote wakeup
signalling for appropriate duration (when LPM is disabled, which is
currently always the case in udc_dwc2). Arbitrarily choose to drive the
resume signalling for 2 ms to have sufficient margin in both directions.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
drivers: udc_dwc2: Add missing GHWCFG bit defines

Add missing GHWCFG2 and GHWCFG3 defines based on nRF54H20 registers
documentation.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -842,9 +847,58 @@ static uint8_t pull_next_ep_from_bitmap(uint32_t *bitmap)
 }
 
+static void dwc2_handle_hibernation_exit(const struct device *dev,
+					 bool remote_wakeup, bool bus_reset)
```
Please use rwup abbreviation.
```
static void dwc2_handle_hibernation_exit(const struct device *dev,
					 const bool rwup, const bool bus_reset)
```

tmon-nordic replied:
Changed, but I don't quite like `rwup`. Saving few characters is not worth all the "what this rwup even means?" that will inevitably happen.

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -130,9 +133,11 @@ struct udc_dwc2_data {
+					  bool remote_wakeup);
 static void dwc2_enter_hibernation(const struct device *dev);
-static void dwc2_exit_hibernation(const struct device *dev);
+static void dwc2_exit_hibernation(const struct device *dev, bool remote_wakeup);
+static void cancel_hibernation_request(struct udc_dwc2_data *const priv);
```
Please get rid of terrible forward definitions.

tmon-nordic replied:
Added new commit reordering the functions prior to all the changes here.

tmon-nordic commented on drivers/usb/udc/Kconfig.dwc2:
```diff
@@ -37,11 +37,3 @@ config UDC_DWC2_THREAD_PRIORITY
 	default 8
```
One k_event is able to handle up to 32 events. For a fully maxed out number of endpoints, there are 32 bits needed - 1 set for transfers queued and 1 set for transfers completed. The third k_event is for the overall.

However, take a look at the actual resource usage. The three k_events do consume less memory and do provide faster handling than the single k_queue. There are absolutely no benefits IMHO for the k_queue and a very serious drawback (losing messages on queue full, which is more likely at High-Speed due to 40x the bandwidth compared to Full-Speed).

jfischer-no replied:
There should be `select EVENTS` between L:9 and L:10

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -924,7 +950,19 @@ static ALWAYS_INLINE void dwc2_thread_handler(void *const arg)
+
+				if (k_event_test(&priv->ep_finished_evt, UINT32_MAX)) {
+					k_event_post(&priv->drv_evt, BIT(DWC2_DRV_EVT_EP_FINISHED));
+				}
```
It is in `dwc2_handle_hibernation_exit()` moved in a later commit in the series (your question already appears as "Outdated")

tmon-nordic committed:
drivers: usb: udc_dwc2: Enter hibernation in thread

Enter hibernation in thread context with the lock held to make sure to
not queue any transfers when the core is hibernated.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
drivers: usb: udc_dwc2: Exit hibernation on disable

The PMU is not reset by core reset and therefore it is necessary to exit
hibernation on DWC2 disable to prevent endless PMU interrupt loop when
the driver is enabled again.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
drivers: usb: udc_dwc2: Remove forward defitions

Reorder functions in order to remove forward function defitions.
No functional changes.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -128,6 +130,7 @@ struct udc_dwc2_data {
 };
 
 static void dwc2_on_bus_reset(const struct device *dev);
+static void dwc2_enter_hibernation(const struct device *dev);
```
Reordered the handlers to get rid of all forward declarations in the file.

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -111,6 +93,12 @@ struct dwc2_reg_backup {
+	/* Transfer triggers (OUT on bits 0-15, IN on bits 16-31) */
+	struct k_event xfer_evt;
+	/* Finished transactions (OUT on bits 0-15, IN on bits 16-31) */
+	struct k_event ep_finished_evt;
```
Renamed.

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -2695,60 +2683,111 @@ static void udc_dwc2_isr_handler(const struct device *dev)
+		}
+	}
+
+	if (evt & BIT(DWC2_DRV_EVT_SETUP)) {
```
Changed.
